python/ray/serve/_private/metrics_utils.py

```diff
@@ -1,138 +1,69 @@
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
 
 
-class _MetricTask:
-    def __init__(self, task_func, interval_s, callback_func):
-        """
-        Args:
-            task_func: a callable that MetricsPusher will try to call in
-                each loop. It should take no arguments, and return a
-                kwargs dictionary to be used to call the callback_func.
-            interval_s: the interval of each task_func is supposed to be called.
-            callback_func: callback function is called when task_func is done, and
-                the result of task_func is passed to callback_func as the first
-                argument, and the timestamp of the call is passed as the second
-                argument.
-        """
-        self.task_func: Callable[[], Dict[str]] = task_func
-        self.interval_s: float = interval_s
-        self.callback_func: Callable[[Any, float]] = callback_func
-        self.last_ref: Optional[ray.ObjectRef] = None
-        self.last_call_succeeded_time: Optional[float] = time.time()
-
-
 class MetricsPusher:
     """
-    Metrics pusher is a background thread that run the registered tasks in a loop.
+    Starts an asyncio task that repeatedly runs registered tasks.
     """
 
-    # If no tasks are registered, sleep for 1s between iterations (until one is
-    # registered).
-    NO_TASKS_REGISTERED_INTERVAL_S = 1
+    def __init__(self, event_loop: asyncio.AbstractEventLoop):
+        self._event_loop = event_loop
+        self._tasks: Dict[str, asyncio.Task] = dict()
+        self._stop_event = asyncio.Event(loop=event_loop)
 
-    def __init__(self):
-        self.tasks: Dict[str, _MetricTask] = dict()
-        self.pusher_thread: Union[threading.Thread, None] = None
-        self.stop_event = threading.Event()
+    def start(self):
+        self._stop_event.clear()
 
     def register_or_update_task(
         self,
         name: str,
         task_func: Callable,
         interval_s: int,
-        process_func: Optional[Callable] = None,
     ) -> None:
         """Register a task under the provided name, or update it.
 
         This method is idempotent - if a task is already registered with
         the specified name, it will update it with the most recent info.
         """
 
-        self.tasks[name] = _MetricTask(task_func, interval_s, process_func)
+        async def new_task():
+            while True:
+                task_func()
+                await async_sleep(interval_s)
 
-    def start(self):
-        """Start a background thread to run the registered tasks in a loop.
+                if self._stop_event.is_set():
+                    return
 
-        We use this background so it will be not blocked by user's code and ensure
-        consistently metrics delivery. Python GIL will ensure that this thread gets
-        fair timeshare to execute and run.
-        """
+        if name in self._tasks:
+            self._tasks[name].cancel()
```

Comment:
The `task_func` cannot currently be `async def`, so this is only able to interrupt the `asyncio.sleep` call. As described above, the graceful way to do this is to instead `await` the stop event. Let's update that logic and remove this.

Reply:
We should probably also include a barrier so that it's not possible for the old task to run after the new task.

Reply:
Perhaps it's better to instead have a dictionary of `task_name -> (sleep_interval, task_func)` and just update that in-place rather than stopping the task and starting a new one.

Reply:
Makes sense! I've changed it to use a dictionary instead, thanks for the suggestion.